### PR TITLE
feat: attitude

### PR DESCRIPTION
**Pull request by HuanCheng65**
## Summary by CodeRabbit

- **New Features**
	- Enhanced user interaction by updating the attitude mechanism for answers and questions.
	- Added support for questions to be associated with groups, allowing for null values.
	- Improved question statistics with detailed counts for answers, comments, and group information.
- **Refactor**
	- Restructured response handling and logic for updating attitudes on answers and questions.
	- Optimized data retrieval in services to include comprehensive details such as attitudes, answer counts, comment counts, and group information.
- **Tests**
	- Expanded end-to-end tests to cover new attitude features and related question and answer statistics.
- **Chores**
	- Updated project settings to adhere to conventional commit guidelines.
- **Documentation**
	- Revised code documentation to accurately reflect the latest functionalities.

### CODE REVIEW

File: src/attitude/DTO/attitude-state.dto.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  user_attitude: AttitudeType;
+
+  constructor(
+    positive_count: number,
+    negative_count: number,
+    user_attitude: AttitudeType,
+  ) {
+    this.positive_count = positive_count;
+    this.negative_count = negative_count;
+    this.difference = positive_count - negative_count;
```
Comment (andylizf) [inline src/attitude/DTO/attitude-state.dto.ts:16]:
```typescript
get difference(): number {
  return this.positive_count - this.negative_count;
}
```

File: src/attitude/attitude.service.ts [hunk 2]
```diff
@@ -37,24 +42,36 @@ export class AttitudeService {
...
-        attitudableId,
-        attitude,
-      },
-    });
+      });
+    } else {
+      await this.prismaService.attitude.delete({
+        where: {
+          attitudableId_userId_attitudableType: {
+            userId,
```
Comment (andylizf) [inline src/attitude/attitude.service.ts:68]: duplicate items
make a variable `attitudeCondition`

**Review by andylizf — CHANGES_REQUESTED** [review]

File: src/questions/DTO/question.dto.ts [hunk 3]
```diff
@@ -11,14 +12,13 @@ export class QuestionDto {
...
-  is_like: boolean;
   my_answer_id: number | null | undefined;
   answer_count: number;
   comment_count: number;
   follow_count: number;
-  like_count: number;
   view_count: number;
   is_group: boolean;
-  group: GroupDto;
+  group?: GroupDto;
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/questions/DTO/question.dto.ts]: https://github.com/SageSeekerSociety/cheese-backend/pull/71/commits/021602d4a50189e3b75ed31d81db02225bbc5df4

File: src/users/users-permission.service.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
+import { Injectable } from '@nestjs/common';
+import { Authorization, AuthorizedAction } from '../auth/auth.service';
+
+@Injectable()
+export class UsersPermissionService {
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/users/users-permission.service.ts:5]: this service can be extended and become part of our permission system

File: src/questions/DTO/question.dto.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/questions/DTO/question.dto.ts]: https://github.com/SageSeekerSociety/cheese-backend/pull/71/commits/0974ac17db9ffcc3cbd26077775eafbb9286720e

File: src/questions/questions.service.ts [hunk 5]
```diff
@@ -221,30 +229,57 @@ export class QuestionsService {
...
-    let user: UserDto | null = null; // For case that user is deleted.
-    try {
-      user = await this.userService.getUserDtoById(
-        question.createdById,
-        viewerId,
-        ip,
-        userAgent,
-      );
-    } catch (e) {
-      // If user is null, it means that one user created this question, but the user
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/questions/questions.service.ts]: https://github.com/SageSeekerSociety/cheese-backend/pull/71/commits/00786669c4f18e0471d82d21687c4a6e1d3888bc

File: src/answer/answer.service.ts [hunk 6]
```diff
@@ -183,31 +200,69 @@ export class AnswerService {
...
+      answerId,
+    );
+    const isFavoritePromise = this.isFavorite(questionId, answerId, viewerId);
+    const groupDtoPromise =
+      answer.groupId == undefined
+        ? Promise.resolve(undefined)
+        : this.groupsService.getGroupDtoById(viewerId, answer.groupId);
+
+    const [
+      ,
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/answer/answer.service.ts:250]: '_' is a good syntax in c#, but it causes a problem here:

```
/home/nictheboy/Documents/cheese-backend/src/answer/answer.service.ts
  250:7  error  '_' is assigned a value but never used  @typescript-eslint/no-unused-vars
```

So, won't fix.

File: src/attitude/DTO/attitude-state.dto.ts (hunk 1 above)
Comment (Nictheboy) [reply src/attitude/DTO/attitude-state.dto.ts:16]: good idea, but when I try this, it doesn't work due to JSON serialization mechanism

File: prisma/schema.prisma [hunk 7]
```diff
@@ -130,29 +130,38 @@ model answer_user_attitude {
...
-  DISLIKE
+  POSITIVE
+  NEGATIVE
 
   @@map("attitude_type")
 }
 
+// Although UNDEFINED is supported,
+// it should not be stored in database.
+enum AttitudeTypeNotUndefined {
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply prisma/schema.prisma:149]: I want to ensure this constraint at database level. This help us while developing.

File: test/question.e2e-spec.ts [hunk 8]
```diff
@@ -607,6 +613,210 @@ describe('Questions Module', () => {
...
+      expect(respond.body.message).toBe(
+        'You have expressed your attitude towards the comment',
+      );
+      expect(respond.statusCode).toBe(201);
+      expect(respond.body.code).toBe(201);
+      expect(respond.body.data.attitudes.positive_count).toBe(1);
+      expect(respond.body.data.attitudes.negative_count).toBe(1);
+      expect(respond.body.data.attitudes.difference).toBe(0);
+      expect(respond.body.data.attitudes.user_attitude).toBe('NEGATIVE');
+    });
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply test/question.e2e-spec.ts]: fixed: https://github.com/SageSeekerSociety/cheese-backend/pull/71/commits/56c9db1ab6791b70846c0f12269c78eadf7fd590

File: src/attitude/attitude.enum.ts [hunk 9]
```diff
@@ -14,14 +14,10 @@ export function parseAttitude(attitude: string): AttitudeType {
...
-    case 'DISLIKE':
-      return AttitudeType.DISLIKE;
-    case 'AGREE':
-      return AttitudeType.AGREE;
-    case 'DISAGREE':
-      return AttitudeType.DISAGREE;
+    case 'POSITIVE':
+      return AttitudeType.POSITIVE;
+    case 'NEGATIVE':
+      return AttitudeType.NEGATIVE;
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/attitude/attitude.enum.ts:20]: This 'LIKE' is used to test InvalidAttitudeTypeError

won't fix

**Review by Nictheboy — APPROVED** [review]

**Review by andylizf — APPROVED** [review]

**Pull request merged**